odl/discr/grid.py

```diff
@@ -1111,8 +1111,8 @@ def uniform_grid_fromintv(intv_prod, shape, nodes_on_bdry=True):
 
     shape = normalized_scalar_param_list(shape, intv_prod.ndim, safe_int_conv)
 
-    if np.shape(nodes_on_bdry) == ():
-        nodes_on_bdry = ([(bool(nodes_on_bdry), bool(nodes_on_bdry))] *
```

Comment:
There should be a type hinting for the `nodes_on_bdry `argument

Reply:
I agree, but am unsure how best to do it &ndash; as I often am in Python, where there is no unambiguous "correct type".

A direct version would be
```
def uniform_grid(min_pt, max_pt, shape,
        nodes_on_bdry: Union[bool, List[Union[bool, Tuple[bool,bool]]]] =True):
    ...
```
That expresses fairly well the different ways the function can be called, but I reckon it looks daunting to the user. An alternative could be to define this union type separately:
```
BoundarySpecification = Union[bool, List[Union[bool, Tuple[bool,bool]]]]

def uniform_grid(min_pt, max_pt, shape, nodes_on_bdry: BoundarySpecification =True):
    ...
```
This has the advantage that the type signature remains short, though it may also be less useful. And arguably, if we make a definition at all then it would be consequent to make it a dedicated _class_, though that seems overkill here.

It is also debatable whether `List` is the right thing. The documentation says "sequence". I a way tuples would be more natural (both because they're also used for array shapes, and because the type of each element can be different and lists are more commonly understood as homogeneous containers), but I find
```
def uniform_grid(min_pt, max_pt, shape,
        nodes_on_bdry: Union[bool, Tuple[Union[bool, Tuple[bool,bool]], ...]] =True):
    ...
```
more confusing than the version with `List`. I'm also not a fan of `Iterable[...]`, that seems to suggest passing a lazily-generated sequence.

To be fair, the Sphinx documentation is actually pretty clear about what the `nodes_on_bdry` does and how it can be specified. Maybe a type hint causes more confusion than it clears up, and I should rather add some comments explaining what happens in the `if isinstance(nodes_on_bdry, bool):` and following lines?

Reply:
Here is how I understand the nodes_on_bdry argument: `nodes_on_bdry : Union[bool, Tuple[bool]]`. If a bool is provided, it sets all boundaries. If a tuple is provided it functions as follows:

1. check that the dimension of the tuple matches the dimension of the array: a`ssert len(nodes_on_bdry) == array.dim`
2. We then have an array describing:

- `nodes_on_bdry = (node_on_bdry_left,node_on_bdry_right)` in dimension 1
- `nodes_on_bdry = (node_on_bdry_left,node_on_bdry_right, node_on_bdry_top, node_on_bdry_bottom)` in dimension 2
- `nodes_on_bdry = (node_on_bdry_left,node_on_bdry_right, node_on_bdry_top, node_on_bdry_bottom, node_on_bdry_front, node_on_bdry_back)` in dimension 3

What do you think about that? :)

Reply:
@Emvlt that's not how it currently works. Your examples should actually be written

- `nodes_on_bdry = ((node_on_bdry_left,node_on_bdry_right),)` in dimension 1
- `nodes_on_bdry = ((node_on_bdry_left,node_on_bdry_right), (node_on_bdry_top, node_on_bdry_bottom))` in dimension 2
- `nodes_on_bdry = ((node_on_bdry_left,node_on_bdry_right), (node_on_bdry_top, node_on_bdry_bottom), (node_on_bdry_front, node_on_bdry_back))`
```
So for example `nodes_on_bdry = ((True, False), (True, True))` for a 2D domain that includes the boundary everywhere except in the right boundary. If that were all, the type hint could be written as
```
Tuple[Tuple[Bool,Bool], ...]
```
or alternatively and IMO clearer
```
List[Tuple[Bool,Bool]]
```
However, the specification can also be shortened to `((True, False), True)`. Or it could be `(True, (False, True))` which means something different: all boundaries included except the top one. This makes it all more concise but leads to much messier types with a union nested inside a list.

I'm not sure if anybody really need this flexibility just to make their user code a few characters shorter. We could use a simplified type hint and then just remark in the docs that it can also be shortened by collapsing `(False, False)` to `False`.

Reply:
I don't think we will come to a good decision here and then, let's leave the interface as it is for now. But I opened [an issue to discuss the use of type hints](https://github.com/odlgroup/odl/issues/1654) for the future.